File: server/pkg/api/sys_overviews.go
package api

const (
	SysOverviewsQueryTypeBySectorID = "bySectorId"
	SysOverviewsQueryTypeByCoords   = "byCoords"
)

type SysOverviewsQueryBySectorID struct {
	SectorID string `json:"sectorId"`
	Limit    int    `json:"limit"`
}

type SysOverviewsQueryByCoords struct {
	RMin       float64 `json:"rMin"`
	RMax       float64 `json:"rMax"`
	ThetaStart float64 `json:"thStart"`
	ThetaEnd   float64 `json:"thEnd"`
	Limit      int     `json:"limit"`
}

type SysOverviewsTableRow struct {
	SystemID    string  `json:"systemId"`
	CoordsR     float64 `json:"coordsR"`
	CoordsH     float64 `json:"coordsH"`
	CoordsTheta float64 `json:"coordsTh"`
	IsExplored  bool    `json:"isExplored"`
	NPlanets    int     `json:"nPlanets"`
	NMoons      int     `json:"nMoons"`
	NAsteroids  int     `json:"nAsteroids"`
	NPops       int     `json:"nPops"`
	NBases      int     `json:"nBases"`
	NCities     int     `json:"nCities"`

	Stars []SysOverviewsTableRowStar `json:"stars"`
}

type SysOverviewsTableRowStar struct {
	ID             string  `json:"starId"`
	TempK          float64 `json:"tempK"`
	LuminositySuns float64 `json:"lumSuns"`
	RadiusAu       float64 `json:"radiusAu"`
	MassSuns       float64 `json:"massSuns"`
	AgeByrs        float64 `json:"ageByrs"`
}

File: ui/src/store/datafront/sys_overviews.ts
import { type StarSystemOverview } from '../../domain/StarSystem';
import { createDatafrontTable } from '../../lib/datafront/table';
import {
    type SysOverviewsQueryByCoords,
    type SysOverviewsQueryBySectorID,
    type SysOverviewsTableRow,
} from '../../lib/net/types.generated';
import { ws } from '../../lib/net/ws';
import { updater, cleaner } from './misc';

export const dfSysOverviews = createDatafrontTable<SysOverviewsTableRow, StarSystemOverview>({
    name: 'sys_overviews',
    ws,
    updater,
    cleaner,
    map: (data) => {
        return {
            id: data.systemId,
            coords: {
                r: data.coordsR,
                h: data.coordsH,
                theta: data.coordsTh,
            },
            isExplored: data.isExplored,
            nAsteroids: data.nAsteroids,
            nPlanets: data.nPlanets,
            stars: data.stars.map((s) => ({
                id: s.starId,
                ageBillionYears: s.ageByrs,
                luminositySuns: s.lumSuns,
                massSuns: s.massSuns,
                radiusAu: s.radiusAu,
                tempK: s.tempK,
            })),
        };
    },
});

export const dfSysOverviewsBySectorId = dfSysOverviews.createQuery<SysOverviewsQueryBySectorID>(
    'sys_overviews/bySectorId',
    (p) => [p.sectorId, p.limit].join(':'),
);

export const dfSysOverviewsByCoords = dfSysOverviews.createQuery<SysOverviewsQueryByCoords>('sys_overviews/byCoords');
